crates/oxc_linter/src/lib.rs

```diff
@@ -49,16 +56,23 @@ impl Linter {
     }
 
     #[must_use]
-    pub fn run(&self, semantic: &Rc<Semantic>) -> Vec<Error> {
-        let ctx = LintContext::new(semantic.clone());
+    pub fn run(&self, semantic: &Rc<Semantic>, source_text: &str) -> LintRunResult {
+        let mut ctx = LintContext::new(semantic.clone());
 
         for node in semantic.nodes().iter() {
             for rule in &self.rules {
-                rule.run(node, &ctx);
+                if let Some(fix) = rule.run(node, &ctx) {
+                    ctx.push_fix(fix);
+                }
```

Comment:
We could probably handle this internally in `Rule::run`, similar to how the diagnostic is reported internally. Thoughts?

Reply:
I prefer handle fixes internally in `Rule::run` such that the rules without fix do not need to return `None`. Only the rules which have fix need to handle this.

Reply:
That makes sense, I was initially worried that providing a mutable `LintContext` to the rule seemed weird because then it could mutate `LintContext` however it wants, but we do the same for diagnostics so it makes sense to align them I think

Reply:
Agree, let's move this into `Rule::run`, the `mut` problem should be handled by interior immutability.

Reply:
Yeah we can move to interior mutability here too. The issue is `LintContext::into_diagnostics` takes ownership of self, so should we introduce a struct that holds the error and the fix:

```rust
pub struct LintDiagnostic<T: Into<Error> = Error> {
  error: T,
  fix: Option<Fix>,
}
```

that way when it comes to fixing, we can pull out the inner value of the `RefCell` of all errors paired w/ their (optional) fixes?

Reply:
What about `LintRunResult::from(ctx)` or something similar? We can remove `into_diagnostics`.

Reply:
Your proposed `LintDiagnostic` has two drawbacks: larger struct size (unnecessary allocation) and larger compile size (monomorphization).

Reply:
> What about LintRunResult::from(ctx) or something similar? We can remove into_diagnostics.

This sounds good. The other approach I looked at is changing `into_diagnostics()` to return `(Vec<Fix>, Vec<Error>)` to avoid the ownership issue, but I think your suggestion sounds better.